### ISSUE TITLE
PR: add different resize and noise-adding policies to input data

### ISSUE THREAD

**Issue by ducha-aiki** [BVLC/caffe#1550]
Is anyone interested in PR, which allows to flexible set different resize (warp, aspect-keeping, different padding, random rotation etc.) and noise policies each with given probability? 

The prototxt looks like following

```
transform_param {
    crop_size: 227
    resize_param {
      prob: 0.5
      resize_mode: WARP
      height: 256
      width: 256
      interp_mode: LINEAR
    }
    resize_param {
      prob: 0.25
      resize_mode: BY_LARGE_SIZE
      height: 256
      width: 256
      interp_mode: AREA
    }
      resize_param {
      prob: 0.2
      resize_mode: BY_SMALL_SIZE_AND_PAD
      height: 256
      width: 256
      pad_mode: MIRRORED
      interp_mode: NEAREST
    }
      resize_param {
      prob: 0.05
      resize_mode: BY_SMALL_SIZE_AND_PAD
      height: 256
      width: 256
      pad_mode: CONSTANT
      pad_value: 104
      pad_value: 117
      pad_value: 124
      interp_mode: NEAREST
    }
    noise_param {
    prob: 0.6
    }
    noise_param {
    hist_eq: true
    prob: 0.2
    }  
    noise_param {
    decolorize: true
    gauss_blur: true
    prob: 0.2
    }  
```

**Comment by RalphMao** [comment]
Thanks!

**Comment by ducha-aiki** [comment]
Done in https://github.com/ducha-aiki/caffe/tree/more-transforms (works only for cv::Mat version of transform)

I plan to clean it up and enhance after #1900 would be merged.

**Comment by RalphMao** [comment]
Can you please offer a prototxt example? The original prototxtx meets the error below:
Error parsing text-format caffe.NetParameter: 12:19: Message type "caffe.LayerParameter" has no field named "transform_param".

**Comment by ducha-aiki** [comment]
@RalphMao are you using my branch or caffe-master? And it is not net parameter, you just need to replace default transform parameter in caffenet proto.

**Comment by RalphMao** [comment]
@ducha-aiki I'm using your branch, and I am still not sure which one should I modify.
Here is part of my prototxt file, could you help point out where is wrong?
layers {
  name: "data"
  type: DATA
  top: "data"
  top: "label"
  data_param{
     source: "examples/imagenet/ilsvrc12_train_lmdb"
    backend: LMDB
    batch_size: 20
  }
  transform_param {
    crop_size: 224
    mean_file: "data/ilsvrc12/imagenet_mean.binaryproto"
    resize_param {
      prob: 0.5
      resize_mode: WARP
      height: 256
      width: 256
      interp_mode: LINEAR
    }
    resize_param {
      prob: 0.25
.....
}
}

**Comment by ducha-aiki** [comment]
It is because you use old V1 interface ("layers" instead of "layer" and type: DATA instead of type: "Data"). 
Correct one :
layer{
name: "data"
type: "Data"
top: "data"
top: "label"
data_param{
source: "examples/imagenet/ilsvrc12_train_lmdb"
backend: LMDB
batch_size: 20
}
    transform_param {
      scale: 0.0039215684
      mirror: true
      crop_size: 227
      mean_value: 114
      resize_param {
        prob: 1
        resize_mode: WARP
        height: 256
        width: 256
        pad_mode: MIRRORED
        pad_value: 114
        interp_mode: LINEAR
        center_object: false
        max_angle: 0
      }
      vertical_mirror: true
      random_90_deg_rot: true
    }
}

And change all other layers (use ca use upgrade_proto binary from tools)

**Comment by RalphMao** [comment]
@ducha-aiki  thanks!

**Comment by immars** [comment]
@ducha-aiki thanks, and here's a gaussian noise implementation based on your code: https://github.com/immars/caffe/tree/aiki-transform

**Comment by ducha-aiki** [comment]
@immars thanks for adding and rebasing! 
PR https://github.com/BVLC/caffe/pull/2252

**Comment by MicahChambers** [comment]
I've been thinking about adding randomized smaller rotations. Since there are already 90 degree rotation, maybe called 'tilt', and make it gaussian?

**Comment by ducha-aiki** [comment]
@MicahChambers you can take then from https://github.com/ducha-aiki/caffe/tree/more-transforms

**Comment by MicahChambers** [comment]
So thats in the proto buff, but not actually implemented yet, correct? 

Eventually all of the data augmentation layers should probably include a parameter for distribution as well. In most cases the center value is far more likely to be the value seen in real data, so scattering around a mean is a better training methodology than uniform. I'll see if I can create a PR for random uniform rotation in a couple of days though.

**Comment by ducha-aiki** [comment]
@MicahChambers no, it IS actually implemented in fork I gave you a link in previous post.
Also, it is implemented in other way in https://github.com/ChenglongChen/caffe-windows

**Comment by MicahChambers** [comment]
Ah, why haven't these been merged yet? @ducha-aiki

**Comment by ducha-aiki** [comment]
@MicahChambers you can ask i.e. @bhack or @mtamburrano about this, I am pretty sure, they could tell nice story about merging ;)

**Comment by mtourne** [comment]
@ducha-aiki I could personally make use of that feature. I'll be trying it locally.

**Comment by akilawickey** [comment]
I tried caffe on Jetson TX1 and got following error

ubuntu@tegra-ubuntu:~/SSD_deep_learning/ssd-tx1$ sudo ~ubuntu/SSD_deep_learning/caffe-ssd/build/tools/caffe test --model="test.prototxt" --weights="VGG_VOC0712_SSD_300x300_iter_60000.caffemodel" --iterations="10000" --gpu 0
[sudo] password for ubuntu:
I1219 03:21:37.389559 2046 caffe.cpp:270] Use GPU with device ID 0
I1219 03:21:37.423753 2046 caffe.cpp:274] GPU device name: NVIDIA Tegra X1
[libprotobuf ERROR google/protobuf/text_format.cc:274] Error parsing text-format caffe.NetParameter: 10:18: Message type "caffe.TransformationParameter" has no field named "resize_param".
F1219 03:21:38.238714 2046 upgrade_proto.cpp:88] Check failed: ReadProtoFromTextFile(param_file, param) Failed to parse NetParameter file: test.prototxt
*** Check failure stack trace: ***
@ 0x7f7a456718 google::LogMessage::Fail()
@ 0x7f7a458614 google::LogMessage::SendToLog()
@ 0x7f7a456290 google::LogMessage::Flush()
@ 0x7f7a458eb4 google::LogMessageFatal::~LogMessageFatal()
@ 0x7f7a70537c caffe::ReadNetParamsFromTextFileOrDie()
@ 0x7f7a6b896c caffe::Net<>::Net()
@ 0x409f00 test()
@ 0x408330 main
@ 0x7f7a0088a0 __libc_start_main
Aborted

**Comment by sequae92** [comment]
I am also getting the same error and its getting aborted.

"10:18: Message type "caffe.TransformationParameter" has no field named "resize_param".
F1219 03:21:38.238714 2046 upgrade_proto.cpp:88] Check failed: ReadProtoFromTextFile(param_file, param) Failed to parse NetParameter file: test.prototxt"

Can anyone guide me towards a possible solution to this ?

**Comment by ducha-aiki** [comment]
Very simple - merge this PR manually, since caffe main branch does not have it yet.

**Comment by shelhamer** [comment]
Thanks for extending the set of data augmentations, but I think the endless possibilities for pre-processing are best left to Python layers or Module layer #5294 to avoid accumulating complexity in the core. Although I'm closing this for that reason, thanks for contributing the warping and noise augmentations.

**Comment by ducha-aiki** [comment]
Vanilla Python layer is bad idea, because DataLayer processing is done is parallel stream, therefore doesn`t make overhead and makes GPU busy. But I don`t care anymore, because switched from caffe long time ago.

**Comment by hammadullah125** [comment]
Error parsing text-format caffe.NetParameter: 16:18: Message type "caffe.TransformationParameter" has no field named "resize_param".

I'm getting this error and unfortunately couldn't find any solution. Any help or hint will be really helpful.

**Comment by ducha-aiki** [comment]
@hammadullah125  because this PR was not merged

**Comment by hammadullah125** [comment]
@ducha-aiki can you please explain this a little bit? I'm very beginner in this field. I would really appreciate your help.

**Comment by ducha-aiki** [comment]
@hammadullah125 
there is no such functionality in caffe framework. To make it work, you would need to first get original caffe, and second merge it with https://github.com/ducha-aiki/caffe/tree/more-transforms
This process is non-trivial, so I don`t recommend you to do it. Instead, better use newer frameworks like pytorch or keras.

**Comment by wucj123** [comment]
> "Error parsing text-format caffe.NetParameter: 16:18: Message type "caffe.TransformationParameter" has no field named "resize_param".
> 
> I'm getting this error and unfortunately couldn't find any solution. Any help or hint will be really helpful.

i have the same problem, have you solved?
